src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h

```diff
@@ -455,6 +455,102 @@ class Simplex_tree_skeleton_simplex_iterator : public boost::iterator_facade<
   int curr_dim_;
 };
 
+/** \brief Iterator over the simplices of the simplicial complex that match the dimension specified by the parameter.
+ *
+ * Forward iterator, value_type is SimplexTree::Simplex_handle.*/
+template<class SimplexTree>
+class Simplex_tree_dimension_simplex_iterator : public boost::iterator_facade<
```

Comment:
I feel like `Simplex_tree_complex_simplex_iterator`, `Simplex_tree_skeleton_simplex_iterator` and `Simplex_tree_dimension_simplex_iterator` are more or less the same class with just different stopping conditions. I am wondering if they would be a clean way to factorize them? This would also highlight more the difference between them (codewise).

Reply:
I agree that it would be nice to factorize them. I think we can introduce the new interface with duplicated code now and delay any refactoring, but if you disagree we can block the PR until it is done.

(it is a shame that C++ coroutines aren't better, it would be so much more readable to implement this with for_each_simplex + yield)